### DISCUSSION TITLE
Automatic link between Id property and identifier for UI controls

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#6119] [Language Ideas]
In WinForms and WebForms, `Name` and `ID` properties, respectively, are also used to generate a field to point to the control in question. I'm trying to use C# (obviously since WebForms is deprecated) and find it annoying to type things like `this.btn = new Button { Id = "btn" }`. I was wondering if there is a way to simplify this. 

One approach would be to automatically generate a field in the containing class, using a source generator. This may work - I plan to look into that.
**Edit:** Having looked at it, it doesn't like it would work. It's aimed at implementing stuff (like partial methods) which you can reference already, but it doesn't like it's able to auto-generate new identifiers. (Please correct me if I'm wrong.)

Another approach would be to create an attribute like so:
```cs
class Control
{
  [SetFromNameOfAssignedField]
  public string? Id { get; set; }
}
// elsewhere
this.ctlAbc = new Control();
// is compiled to
this.ctlAbc = new Control { Id = "ctlAbc", };
```
Does anybody have thoughts on this? Any other ideas I didn't think of?

**Comment by HaloFour** [comment]
How often do you need to manually create these controls and have their IDs match their identifiers vs. using razor or whatever designer to do that for you?  It's been a while since I've done web UX, and when I've used code to manually construct bits of UI and it was pretty rare that the ID I assigned to the control matched the identifier to which I assigned that control, if any.

**Reply by TahirAhmadov** [reply]
I purposefully avoid Razor and MVC in general, in favor of a different approach. There are reasons for that - I could elaborate, but it's not that important for this question.
And, my question also applies to desktop development, there, it's controls everywhere.
The identifier doesn't have to match the ID perfectly - although it's helpful from readability perspective. The main objective is to actually have an identifier created in the first place - to reference the controls for data binding, presentation manipulation, etc.

**Reply by HaloFour** [reply]
Considering that this has been the design of UX development since C# and .NET 1.0 and your friction comes specifically from avoiding the tools offered that make said development simpler I don't think the language should be going out of its way to add features specifically for this case.  In most cases those IDs don't matter anyway.

Also, if the feature is adopted as suggested and the UX frameworks update controls with such an annotation that would change the meaning of existing code.

**Reply by TahirAhmadov** [reply]
> Considering that this has been the design of UX development since C# and .NET 1.0 and your friction comes specifically from avoiding the tools offered that make said development simpler I don't think the language should be going out of its way to add features specifically for this case. 

What has been the design of UX since C# and .NET 1.0? Razor showed up much later. If you're talking about controls, then yes - they have been around forever - and that's what I'm trying to use. I'm confused as to what you're saying here.
If your point is that "not using Razor is not a supported scenario in C#", I would strongly disagree. 1, Razor is not C#; it's a different language - which is related to C#. 2, if Razor is the **solution** for Web UI, then why expose the middlewares , etc.? Why not just let people create Razor pages and be done with it? Answer - specifically because Razor is just one of many options on how to create a web site using C# and .NET.

> In most cases those IDs don't matter anyway.

They do, a lot. First of all, in both Win and WebForms, descriptive names (and identifiers) contribute to readable code. Secondly, on the Web, the IDs are also often used by JavaScript for client-side manipulation; it would be weird if C# referred to the same control as `okButton` and JavaScript as `potatoButton`.

> Also, if the feature is adopted as suggested and the UX frameworks update controls with such an annotation that would change the meaning of existing code.

Yes, if a language feature is added which introduces a new behavior, and a library changes its code to use the language feature, then code using that library will have to be adapted to expect the new behavior, yes. We're talking about a few years away from today. I'm not sure what the problem is. 
Besides, I just proposed a very simple attribute; it doesn't have to be hard and fast. For example, it can have an option to only set the property if it's null; or only if it's not set in the init-block as part of the assignment to that field; etc.

**Reply by HaloFour** [reply]
> What has been the design of UX since C# and .NET 1.0?

Controls having an `ID` property which the designer automatically keeps in sync with the field name.  That's been true of the WinForm and WebForm designers since Visual Studio 2002.

> Yes, if a language feature is added which introduces a new behavior, and a library changes its code to use the language feature, then code using that library will have to be adapted to expect the new behavior, yes. We're talking about a few years away from today. I'm not sure what the problem is.

There's a ton of existing code that targets those UX frameworks, all of which may be silently and unintentionally impacted if the language and UX framework adopt this proposal.  That's unacceptable to me.  If I left the ID field blank I did so intentionally and don't want the language to assume it can guess what it should put there, especially since duplicates could lead to breaks and the language could never tell what all of the rest of the code interacting with that UX framework might be doing.  If you're working with a helper class that wraps the control the resulting behavior could be completely wrong, resulting in duplicated IDs and very odd behavior at runtime.

IMHO, if you're intentionally avoiding the easy way to do UX design, you accept that it will come with some additional friction.

**Reply by HaloFour** [reply]
You added locals to the mix, and I have plenty of code that will create controls in a loop.  I definitely don't want any of those controls to silently inherit their ID from the name of the local I happened to use, especially since that will definitely result in duplicate IDs.

**Reply by TahirAhmadov** [reply]
> > What has been the design of UX since C# and .NET 1.0?
> 
> Controls having an `ID` property which the designer automatically keeps in sync with the field name. That's been true of the WinForm and WebForm designers since Visual Studio 2002.

Oh yes, I'm not aiming to change WinForms; the designer already does that. However, other desktop UI frameworks may be developed by third party vendors. Also, WebForms is deprecated, so it will **have** to be a third party framework for somebody (like me) who prefers control-based approach, and developing a designer is out of the question for me and most small teams out there.

> There's a ton of existing code that targets those UX frameworks, all of which may be silently and unintentionally impacted if the language and UX framework adopt this proposal. That's unacceptable to me. If I left the ID field blank I did so intentionally and don't want the language to assume it can guess what it should put there, especially since duplicates could lead to breaks and the language could never tell what all of the rest of the code interacting with that UX framework might be doing. If you're working with a helper class that wraps the control the resulting behavior could be completely wrong, resulting in duplicated IDs and very odd behavior at runtime.
> 
> IMHO, if you're intentionally avoiding the easy way to do UX design, you accept that it will come with some additional friction.

It doesn't have to be silently; this can be controlled by a flag of some sort, like an opt-in compilation symbol or a static boolean flag. All these are implementation details.

> You added locals to the mix, and I have plenty of code that will create controls in a loop. I definitely don't want any of those controls to silently inherit their ID from the name of the local I happened to use, especially since that will definitely result in duplicate IDs.

I meant instantiation, not just assignment. And my point was fields all along - locals were a typo - and I see what you mean, it wouldn't work in scenarios like loops.

**Reply by HaloFour** [reply]
> It doesn't have to be silently; this can be controlled by a flag of some sort, like an opt-in compilation symbol or a static boolean flag. All these are implementation details.

That would create a dialect of the language.

**Reply by HaloFour** [reply]
> I meant instantiation, not just assignment.

That's what those loops are doing, instantiating new controls in a loop based on some data source.

**Reply by CyrusNajmabadi** [reply]
To Halo's point.  Our default position on things like flags is a flat out 'no'.  The only exceptions we're willing to take on this are for massive, ecosystem-wide, enormous benefits that basically have to demonstrate insane value.  To give an example of that: NRT.  That's how impactful and important the lang feature needs to be to warrant a flag.

**Reply by TahirAhmadov** [reply]
> > It doesn't have to be silently; this can be controlled by a flag of some sort, like an opt-in compilation symbol or a static boolean flag. All these are implementation details.
> 
> That would create a dialect of the language.

My question is, why is this such a big problem? No framework out there is mandated to adopt this behavior. If you maintain a framework which has N users, and you modify an existing property to do this, it would be a decision you make; in terms of user outreach, demand/feedback from users, typical use cases, etc. - that's a question for you, the framework maintainer - not the language designers.

**Reply by TahirAhmadov** [reply]
> > I meant instantiation, not just assignment.
> 
> That's what those loops are doing, instantiating new controls in a loop based on some data source.

Sure, still - my original question was fields, not locals.

**Reply by HaloFour** [reply]
@TahirAhmadov 

> My question is, why is this such a big problem?

For every flag the permutations of flavors of the language effectively doubles, and every combination of those flags needs to be tested and supported.  NRTs was found to be a high enough bar to create a dialect, and even in that case the language team was very careful to ensure that it would cause no changes to the runtime behavior of the code.

> If you maintain a framework which has N users

Language features have to be designed completely independently of how it may be used by any given framework.

**Reply by TahirAhmadov** [reply]
> To Halo's point. Our default position on things like flags is a flat out 'no'. The only exceptions we're willing to take on this are for massive, ecosystem-wide, enormous benefits that basically have to demonstrate insane value. To give an example of that: NRT. That's how impactful and important the lang feature needs to be to warrant a flag.

This whole conversation started out of (IMO) the mistaken idea that all UI frameworks out there are mandated to go and immediately modify existing properties by decorating them with this attribute. I never said; my intention is to use it for new development in my projects. Existing frameworks can do what they see fit. For example, they can add another property; for example, if `Id` already exists, they can add `Name` and slowly migrate to it; or they can elect to do nothing at all.

**Reply by HaloFour** [reply]
> This whole conversation started out of (IMO) the mistaken idea that all UI frameworks out there are mandated to go and immediately modify existing properties by decorating them with this attribute

Your use case was the existing UX frameworks, if there is no intent for any of them to adopt this change then I see no point at all in the language adding such a feature.

**Reply by TahirAhmadov** [reply]
> > My question is, why is this such a big problem?
> 
> For every flag the permutations of flavors of the language effectively doubles, and every combination of those flags needs to be tested and supported. NRTs was found to be a high enough bar to create a dialect, and even in that case the language team was very careful to ensure that it would cause no changes to the runtime behavior of the code.

You completely misunderstood my question. I wasn't asking about a language flag. My question was, why is this such a big problem whether a framework chooses to adopt a new language feature and it how it affects its users in the future? 

> > If you maintain a framework which has N users
> 
> Language features have to be designed completely independently of how it may be used by any given framework.

Exactly! And I'm baffled by your insistence that we have to try to confront future issues caused by a framework maintainer, by preventing language improvement.

**Reply by CyrusNajmabadi** [reply]
![image](https://user-images.githubusercontent.com/4564579/168141792-9105b6bd-da6b-48bf-a1a8-53321fe4efb0.png)

**Reply by TahirAhmadov** [reply]
> Your use case was the existing UX frameworks, if there is no intent for any of them to adopt this change then I see no point at all in the language adding such a feature.

Not at all. My point was, WinForms designer already does that. WebForms ASPX parser did that, too, but WebForms is deprecated. I'm trying to make a new custom control-based UI framework, because it's preferable for me than using Razor, for reasons that are irrelevant here. Also, there are other UI frameworks third-party vendors may choose to develop (and already develop). Nowhere did I say that any existing framework is mandated to modify any of its properties. (They may choose to do so, per their own decision making process.)

**Reply by HaloFour** [reply]
@TahirAhmadov 

> My question was, why is this such a big problem whether a framework chooses to adopt a new language feature and it how it affects its users in the future?

Because a language change that affects the meaning of already legal code should have an exceptionally high bar.

> And I'm baffled by your insistence that we have to try to confront future issues caused by a framework maintainer, by preventing language improvement.

It's not caused by the framework maintainer, it's caused by the language.  And if this is not a proposal intended to help with UX design with existing frameworks like WinForms, WPF, MVC, etc., then I don't see why it should exist at all.

**Reply by HaloFour** [reply]
> I'm trying to make a new custom control-based UI framework,

So you want a language change to benefit a single user, you.

**Reply by TahirAhmadov** [reply]
> ![image](https://user-images.githubusercontent.com/4564579/168141792-9105b6bd-da6b-48bf-a1a8-53321fe4efb0.png)

Good job cutting out the rest of my response. Here it is:

> My question is, why is this such a big problem? No framework out there is mandated to adopt this behavior. If you maintain a framework which has N users, and you modify an existing property to do this, it would be a decision you make; in terms of user outreach, demand/feedback from users, typical use cases, etc. - that's a question for you, the framework maintainer - not the language designers.

It is absolutely clear that I was making a different point and responding to something else - the mistaken notion that all existing frameworks are automatically impacted. You two seem determined to derail the conversation by twisting my words - by purposefully mis-interpreting my reply and cutting my words out of context.

I'm actually frustrated by the nonconstructive behavior you are demonstrating. Perhaps you can reprimand yourself, in accordance with your rules on constructive behavior.

**Reply by TahirAhmadov** [reply]
> Because a language change that affects the meaning of already legal code should have an exceptionally high bar.

It doesn't. Did you read anything I wrote? I proposed adding an attribute, which would then need to be decorated to a property. What did you think I meant? That the language auto-decorates all properties named "Id" on all classes named "Control"? ....

> It's not caused by the framework maintainer, it's caused by the language. And if this is not a proposal intended to help with UX design with existing frameworks like WinForms, WPF, MVC, etc., then I don't see why it should exist at all.

Completely wrong. Language adds an attribute (or rather, starts handling a new attribute), and a framework chooses whether to use the attribute. Again, please read what I wrote before replying. Otherwise, it looks a lot like straw-manning.

**Reply by HaloFour** [reply]
Finding feedback frustrating is not an indicator of any violation of the code of conduct.  Any and every proposal must be able to answer tough questions, starting with "why?".

**Reply by TahirAhmadov** [reply]
> So you want a language change to benefit a single user, you.

Aren't 100% of suggested features raised by developers who encounter the need for such a feature? 
Also, how do you know other frameworks will not adopt this feature in the future, with appropriate safeguard processes in place, such as communicating the breaking change to their respective user bases?

This message of your above - that I'm demanding something for myself only - is outright personally vindictive. What is wrong with you? Am I your enemy? I sense hatred and/or disdain from you lot in this discussion.

**Reply by TahirAhmadov** [reply]
> Finding feedback frustrating is not an indicator of any violation of the code of conduct. Any and every proposal must be able to answer tough questions, starting with "why?".

Feedback is not the same as literally cutting my words out of context. "Cutting words out of context" is a form of lying.

**Reply by HaloFour** [reply]
> I proposed adding an attribute, which would then need to be decorated to a property.

Which would change the behavior of existing code consuming that property.

> Completely wrong. Language adds an attribute (or rather, starts handling a new attribute), and a framework chooses whether to use the attribute. Again, please read what I wrote before replying. Otherwise, it looks a lot like straw-manning.

There's two logical conclusions here, either it applies to the existing UX frameworks in which case billions of lines of existing code could potentially be impacted, or it only applies to this framework you happen to be writing, in which case it benefits virtually noone.  Neither of those cases justify a language change in my opinion.

**Reply by HaloFour** [reply]
Your words were not taken out of context.  The _language_ would need to offer a flag, creating a dialect.  It doesn't matter that it's optional for a framework to actually adopt the attribute.

**Reply by TahirAhmadov** [reply]
> There's two logical conclusions here, either it applies to the existing UX frameworks in which case billions of lines of existing code could potentially be impacted, or it only applies to this framework you happen to be writing, in which case it benefits virtually noone. Neither of those cases justify a language change in my opinion.

Nope. There is a third option - some new frameworks, be it mine or others', choose to use this feature; others don't. Some existing frameworks choose to use this feature; others don't. 
If a framework you use starts using this feature, and does so in a way that introduces breaking changes, you can take it up with that framework.
The above two statements are true for many, if not all, language and BCL additions.

**Reply by CyrusNajmabadi** [reply]
This is the context of that question. Adding a flag would create a dialect and that has that high bar to get over.  The later part of your post doesn't change that.  The crux of hte matter is that to go this route we'd need such exceptional value to consider it.

> Aren't 100% of suggested features raised by developers who encounter the need for such a feature?

Yes, and if it's that narrow, we normally don't take it.  Consider that we only take a small percentage of proposals.  It is the case that this is often the reason they don't move forward.  We far prefer taking proposals with either broad value to the entire ecosystem, or exceptional value to a subset  that is very important.

**Reply by HaloFour** [reply]
> some new frameworks, be it mine or others', choose to use this feature; others don't.

Which makes it a solution in search of a problem.

You seem to continue changing the nature of this proposal in order to dodge the problematic implications that come with it, but in doing so you also make it substantially less relevant.

**Reply by TahirAhmadov** [reply]
> Your words were not taken out of context. The _language_ would need to offer a flag, creating a dialect. It doesn't matter that it's optional for a framework to actually adopt the attribute.

No it wouldn't. The framework can just not use the attribute and all is well. Or, they can add a new property which has this behavior, not to modify their existing properties.
In the real world, UI frameworks come and go so often, that given that this change (if done) will be done in a couple of years, by that point there will be new frameworks in development - and for them, it would not be a breaking change.
And the most resilient of the .NET UI frameworks - WinForms - does not need this, because it's designer takes care of it already.
All of which brings me back to my question, why is this - un-thought through adoption by third-party frameworks - such a big problem?

**Reply by HaloFour** [reply]
> In the real world, UI frameworks come and go so often, that given that this change (if done) will be done in a couple of years, by that point there will be new frameworks in development - and for them, it would not be a breaking change.

I don't see the point in adding a language feature to target fly-by-night UX frameworks all of which will be replaced within a year when none of those UX frameworks will approach a single percentage point of adoption of the primary UX frameworks that already exist and could never adopt this feature.  This _still_ sounds much too narrowly scoped to benefit only whatever it is that you happen to be working on right now and nobody else.

**Reply by CyrusNajmabadi** [reply]
> No it wouldn't. The framework can just not use the attribute and all is well. Or, they can add a new property which has this behavior, not to modify their existing properties.

What flag are you referring to then in this statement:

> It doesn't have to be silently; this can be controlled by a flag of some sort, like an opt-in compilation symbol or a static boolean flag. All these are implementation details.

**Reply by TahirAhmadov** [reply]
> This is the context of that question. Adding a flag would create a dialect and that has that high bar to get over. The later part of your post doesn't change that. The crux of hte matter is that to go this route we'd need such exceptional value to consider it.

I mentioned a flag in the middle of a back and forth as a solution to a very special challenge:
1) we add this language feature;
2) a significant existing framework chooses to modify an existing property, thus creating regressions in user code.

I understand that adding "dialects" are a no go; consider my suggestion of it as retracted.

However, I went back and am questioning the #2 premise above. If you read the entirety of my comment, part of which you copied to make me look like a fool, I wasn't talking about the flag. Hence, "taking my words out of context".

> > Aren't 100% of suggested features raised by developers who encounter the need for such a feature?
> 
> Yes, and if it's that narrow, we normally don't take it. Consider that we only take a small percentage of proposals. It is the case that this is often the reason they don't move forward. We far prefer taking proposals with either broad value to the entire ecosystem, or exceptional value to a subset that is very important.

Fine - that's a perfectly acceptable response. "Sorry we don't think there are enough UI frameworks who would use this to justify prioritizing this" - no problem. 
But there was no need to make it look like I was saying something I wasn't.

**Reply by CyrusNajmabadi** [reply]
>  part of which you copied to make me look like a fool

I didn't copy that to make you look like a fool.  I copied that because it was what i was responding to, and in teh context of hte original discussion and with what you wrote afterwards it was still relevant.

**Reply by TahirAhmadov** [reply]
> > No it wouldn't. The framework can just not use the attribute and all is well. Or, they can add a new property which has this behavior, not to modify their existing properties.
> 
> What flag are you referring to then in this statement:
> 
> > It doesn't have to be silently; this can be controlled by a flag of some sort, like an opt-in compilation symbol or a static boolean flag. All these are implementation details.

I did suggest a compilation flag as a possible adoption tool, without thinking twice about it. 
But I later went back (in my post that was mis-quoted) to the question of, whether it's such a big problem to begin with - meaning, there is no need to do anything like a flag to begin with.

**Reply by CyrusNajmabadi** [reply]
> Fine - that's a perfectly acceptable response. 

As was the explanation around flags/dialects/etc.  These all weigh into teh decision making process and i'm transparently stating that as it's relevant in assessing these proposals.  There was no intent in making you look like a fool.  I was just showing you what i was seeing and what i was responding to (and that includes what you wrote both before and after).  This point in particular though i was addressing *specifically* because of it's high relevance to the topic here.  None of it was personal.

**Reply by TahirAhmadov** [reply]
> > In the real world, UI frameworks come and go so often, that given that this change (if done) will be done in a couple of years, by that point there will be new frameworks in development - and for them, it would not be a breaking change.
> 
> I don't see the point in adding a language feature to target fly-by-night UX frameworks all of which will be replaced within a year when none of those UX frameworks will approach a single percentage point of adoption of the primary UX frameworks that already exist and could never adopt this feature. This _still_ sounds much too narrowly scoped to benefit only whatever it is that you happen to be working on right now and nobody else.

1, "fly by night" is relative and vague. For example, Silverlight stayed around for a number of years; it's not "within a year". Also, other frameworks exist and continue being marginally supported, but don't receive many improvements. 
2, even if UI frameworks came and went every 6 months, the option to auto-assign ID property would be potentially beneficial to every one of them going forward. It's not like, a framework is deprecated, and this attribute and the language support for it is deprecated at the same time. 
3. It would be a general feature aimed at automatically assigning the field name to a property; perhaps it can be used for scenarios outside of UI, too. (Granted, I can't think of any at the moment.)

But again, by no means am I claiming that this should be priority #1 for the language this second. It wouldn't even be priority #1 out of the language improvements I would like to see.

**Reply by TahirAhmadov** [reply]
> > Fine - that's a perfectly acceptable response.
> 
> As was the explanation around flags/dialects/etc. These all weigh into teh decision making process and i'm transparently stating that as it's relevant in assessing these proposals. There was no intent in making you look like a fool. I was just showing you what i was seeing and what i was responding to (and that includes what you wrote both before and after). This point in particular though i was addressing _specifically_ because of it's high relevance to the topic here. None of it was personal.

There was probably a misunderstanding here, then. 
Forget the flag; all I'm asking for is, an attribute is treated as, "when a new instance of containing type is assigned to a field, set the property value to that field name".

**Comment by CyrusNajmabadi** [comment]
I'm not seeing anything related to C# here.  What language feature are you proposing?

**Reply by TahirAhmadov** [reply]
I was thinking if we added that attribute, it would silently add a property setter. 
```cs
this.btn = new Button { };
// is compiled to
this.btn = new Button { Id = "btn", };
```
**Correction:** I had fields in mind, not locals. But it could work for locals and properties, too. Think of it as debug caller method name attribute like feature.

**Comment by TahirAhmadov** [comment]
I thought of another way to solve this: after the control tree is constructed, go through fields of control types using reflection, and assign IDs automatically. This logic can be IL emitted and cached into a delegate by "container" type (be it page/form/user control).

**Reply by HaloFour** [reply]
Sounds like something you could accomplish in your own framework without requiring any language features and without affecting any other UX framework.
